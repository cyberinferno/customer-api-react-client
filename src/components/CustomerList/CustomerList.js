/**
 * Customer list component
 */

import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { fetchCustomersStarted } from '../../actions';

class CustomerList extends Component {
  componentDidMount() {
    const { page } = this.props;
    if (page !== null) {
      this.props.fetchCustomersStarted(page);
    } else {
      this.props.fetchCustomersStarted();
    }
  }

  componentDidUpdate() {
    const { page, data } = this.props;
    if (typeof data.pagination !== 'undefined' && page !== null && parseInt(page, 10) !== parseInt(data.pagination.current, 10)) {
      this.props.fetchCustomersStarted(page);
    }
  }

  renderCustomerData() {
    const { data } = this.props;
    if (typeof data.results === 'undefined') {
      return (
        <div className="row">
          <div className="col-sm-8">
            <table className="table table-bordered">
              <thead>
                <tr>
                  <th>#</th>
                  <th>Name</th>
                  <th>Phone Number</th>
                </tr>
              </thead>
            </table>
          </div>
        </div>
      );
    }
    return (
      <div className="row">
        <div className="col-sm-8">
          <table className="table table-bordered">
            <thead>
              <tr>
                <th>#</th>
                <th>Name</th>
                <th>Phone Number</th>
                <th>Actions</th>
              </tr>
            </thead>
            <tbody>
              {data.results.map((item) => (
                <tr key={item.id}>
                  <td>{item.id}</td>
                  <td>{item.name}</td>
                  <td>{item.phone}</td>
                  <td><Link to={`/customer/${item.id}`}>View Details</Link></td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      </div>
    );
  }

  renderPagination() {
    const { loading, error, data } = this.props;
    if (loading || error !== '' || typeof data.pagination === 'undefined') {
      return ('');
    }
    const { pagination } = data;
    const buttons = [];
    if (pagination.current === 1 || pagination.total === 1) {
      buttons.push(<li key="prev" className="page-item disabled"><span className="page-link">Previous</span></li>);
    } else {
      buttons.push(<li key="prev" className="page-item"><Link className="page-link" to={`?page=${pagination.current - 1}`}>Previous</Link></li>);
    }
    for (let i = 1; i <= pagination.total; i++) {
      if (pagination.current === i) {
        buttons.push(<li key={i} className="page-item active"><span className="page-link">{i}</span></li>);
      } else {
        buttons.push(<li key={i} className="page-item"><Link className="page-link" to={`?page=${i}`}>{i}</Link></li>);
      }
    }
    if (pagination.current === pagination.total) {
      buttons.push(<li key="next" className="page-item disabled"><span className="page-link">Next</span></li>);
    } else {
      buttons.push(<li key="next" className="page-item"><Link className="page-link" to={`?page=${pagination.current + 1}`}>Next</Link></li>);
    }
    return (
      <nav>
        <ul className="pagination">
          {buttons}
        </ul>
      </nav>
    );
  }

  render() {
    let content;
    const { loading, error } = this.props;
    if (loading) {
      content = <b>Loading...</b>;
    } else {
      if (error !== '') {
        content = <b>error</b>;
      } else {
        content = this.renderCustomerData();
      }
    }
    return (
      <div>
        {content}
        {this.renderPagination()}
      </div>
    );
  }
}

CustomerList.propTypes = {
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string.isRequired,
  fetchCustomersStarted: PropTypes.func.isRequired,
  data: PropTypes.object.isRequired,
  page: PropTypes.string,
};

const mapStateToProps = (state) => {
  const {
    loading,
    data,
    error,
  } = state.customerList;

  return {
    loading,
    data,
    error,
  };
};

export default connect(
  mapStateToProps,
  {
    fetchCustomersStarted,
  }
)(CustomerList);
