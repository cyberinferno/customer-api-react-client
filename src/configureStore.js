/**
 * Create the store with dynamic reducers
 */

import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import logger from 'redux-logger';

import createReducer from './reducers';
import rootSaga from './sagas';

export default function configureStore(initialState = {}) {
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore(
    createReducer(),
    initialState,
    applyMiddleware(logger, sagaMiddleware)
  );

  sagaMiddleware.run(rootSaga);

  return store;
}
