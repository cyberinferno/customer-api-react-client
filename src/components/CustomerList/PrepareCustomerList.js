/**
 * Returns customer list component by inserting current page number
 */

import React from 'react';
import { useLocation } from 'react-router-dom';

import CustomerList from './CustomerList';

function useQuery() {
  return new URLSearchParams(useLocation().search);
}

function PrepareCustomerList() {
  const query = useQuery();
  return (<CustomerList page={query.get('page')} />);
}

export default PrepareCustomerList;
