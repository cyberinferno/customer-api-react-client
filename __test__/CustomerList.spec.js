/**
 * Customer list component test
 */

import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import CustomerList from '../src/components/CustomerList/CustomerList';

configure({ adapter: new Adapter() });

describe('>>> Customer list test', () => {
  const initialState = {
    customerList: {
      error: '',
      loading: false,
      data: {
        results: [
          {
            id: 1,
            name: 'Customer 1',
            phone: 9876543210,
          },
        ],
        pagination: {},
      },
    },
  };
  const mockStore = configureStore();
  let store;
  let wrapper;

  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <Router>
          <CustomerList page="1" />
        </Router>
      </Provider>
    );
  });

  it('+++ Component gets rendered', () => {
    expect(wrapper.find(CustomerList).length).toEqual(1);
  });

  it('+++ Props match with initialState', () => {
    expect(wrapper.find(CustomerList).prop('error')).toEqual(initialState.error);
    expect(wrapper.find(CustomerList).prop('loading')).toEqual(initialState.loading);
    expect(wrapper.find(CustomerList).prop('data')).toEqual(initialState.data);
  });

  it('+++ Contains table', () => {
    expect(wrapper.find('table').length).toEqual(1);
  });

  it('+++ Contains proper amount of rows', () => {
    expect(wrapper.find('tr').length).toEqual(2);
  });

  it('+++ Table contains the right data', () => {
    expect(wrapper.find('th').get(0).props.children).toBe('#');
    expect(wrapper.find('th').get(1).props.children).toBe('Name');
    expect(wrapper.find('th').get(2).props.children).toBe('Phone Number');
    expect(wrapper.find('th').get(3).props.children).toBe('Actions');
    expect(wrapper.find('td').get(0).props.children).toBe(1);
    expect(wrapper.find('td').get(1).props.children).toBe('Customer 1');
    expect(wrapper.find('td').get(2).props.children).toBe(9876543210);
  });

  it('+++ Capturing Snapshot of CustomerList', () => {
    const renderedValue = renderer.create(
      <Provider store={store}>
        <Router>
          <CustomerList page="1" />
        </Router>
      </Provider>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});
