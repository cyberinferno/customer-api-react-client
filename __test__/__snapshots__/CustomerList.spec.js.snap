// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Customer list test +++ Capturing Snapshot of CustomerList 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-sm-8"
    >
      <table
        className="table table-bordered"
      >
        <thead>
          <tr>
            <th>
              #
            </th>
            <th>
              Name
            </th>
            <th>
              Phone Number
            </th>
            <th>
              Actions
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              1
            </td>
            <td>
              Customer 1
            </td>
            <td>
              9876543210
            </td>
            <td>
              <a
                href="/customer/1"
                onClick={[Function]}
              >
                View Details
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <nav>
    <ul
      className="pagination"
    >
      <li
        className="page-item"
      >
        <a
          className="page-link"
          href="/?page=NaN"
          onClick={[Function]}
        >
          Previous
        </a>
      </li>
      <li
        className="page-item disabled"
      >
        <span
          className="page-link"
        >
          Next
        </span>
      </li>
    </ul>
  </nav>
</div>
`;
